### PR TITLE
[rush] Restore log files when cache hit

### PR DESCRIPTION
**Pull request by chengcyber**
## Summary

This PR add a feature: restoring operation log files when cache hit

## Details

Introduce a new "OperationMetadataManager"

1. when a task successful without cache, save the log files into `.rush/temp/operation/<identifier>/all.log` and `.rush/temp/operation/<identifier>/error.log`.
2. when a task successful with cache hit, append the `all.log` file content to current operation log and copy the error log if exists.

## How it was tested

Tested in rushstack and a repo without phase commands.

1. Save the log file
2023-01-12 at 15 19

2. Append the cached log content to current log
2023-01-12 at 15 21

## Impacted documentation

 Nope

### CODE REVIEW

**Review by dmichon-msft — COMMENTED** [review]
I appreciate the consolidation into OperationMetadataManager. At some point it will be necessary to finally move a lot of logic out of ShellOperationRunner (which should only own the logic to invoke a subprocess, not any of the caching, logging, etc.), but that's not really a blocker.

**Comment by chengcyber** [comment]
Hi @dmichon-msft , code changes applied. Please take another look~

**Comment by elliot-nelson** [comment]
@chengcyber I was wondering, do I need to have any special config to enable the feature?

I pulled branch and test as follows:

```console
$ rush install
$ node ~/dev/rushstack/apps/rush/lib/start-dev.js test --verbose --to <someproject>

( a bunch of jest output)

$ ls -al <someproject>/.rush/temp/operation/_phase_test/
-rw-r--r--  1 elliot.nelson  staff   47 Feb  4 11:50 state.json
```

No log file seems to be created (so I couldn't test the part where it retrieves it from cache.)

**Review by dmichon-msft — APPROVED** [review]

**Comment by chengcyber** [comment]
@elliot-nelson Thanks for the information! Actually, i broke something accidently at that time. Now it has been fixed!
image

File: common/changes/@microsoft/rush/feat-build-cache-log_2023-01-11-16-25.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Restore log files when cache hit",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/feat-build-cache-log_2023-01-11-16-25.json:5]:
```suggestion
      "comment": "Include operation log files in the cache, and restore them during cache hits.",
```

**Review by iclanton — APPROVED** [review]

**Comment by elliot-nelson** [comment]
Thanks @chengcyber! It's working great :+1:

(Worth noting that this feature on its own changes nothing about build output -- `rush build --verbose` still shows _this build's_ information, not the contents of `all.log` -- which is what I expected.)

**Comment by octogonz** [comment]
> (Worth noting that this feature on its own changes nothing about build output -- `rush build --verbose` still shows _this build's_ information, not the contents of `all.log` -- which is what I expected.)

@chengcyber I had the opposite expectation -- shouldn't `rush build --verbose` faithfully reproduce the log outputs from a build that was restored from the cache?

**Pull request merged**